src/preload.py

```diff
@@ -610,7 +610,17 @@ def write_resin_device_pinning(app_data, output):
 
 
 def write_apps_json(data, output):
-    """Writes data dict to output as json"""
+    """Updates or creates preloaded state"""
+    if os.path.isfile(output):
+        with open(output, "r") as f:
+            current_state = json.load(f)
+        for uuid in data.keys():
+            if len(uuid) != len(current_state.keys()[0]):
+                raise Exception("Current and target preloaded states not "
+                                "compatible")
+            if uuid not in current_state['apps']:
```

Comment:
But this would mean that running preload consecutive times (on the same, already-preloaded image) would not overwrite previous entries of `apps.json`, which is not good UX (unexpected behaviour from users' point of view).

If you simply delete this `if` clause and always run `current_state['apps'][uuid] = data[uuid]`, this might work better? I understand we don't want to wipe a pre-existing `apps.json`, but we still want to overwrite entries where the `uuid` matches, don't we?

Reply:
There is a check for the existence of `apps.json` so if it is not there the behaviour is just as it was before.
```
if os.path.isfile(output):
```
I really don't like the idea of the CLI modifying the current state that exactly matches what the OS was built with. It should just add apps but not touch existing apps, specially hostos apps as it might break the system.

Reply:
> Why would you want to overwrite entries?

Here's the scenario:

1. Run `balena preload` and select a certain release "R1" of user application "A1".
1.1 The image file now contains an `apps.json` file that lists the selected release "R1" for app "A1".
2. _"Oh wait! I've preloaded the wrong release, duh. I meant to preload release "R2", not release "R1". Let me run it again."_
3. Run `balena preload` again for the same image file and same app, but this time select release "R2".
3.1 Expected result: the image file (apps.json) now lists release "R2" for app "A1".
3.2 Actual, unexpected result: the image file (apps.json) still lists release "R1" for app "A1", because the code above refuses to overwrite it.

All the above happening before the image is flashed, which means we were not on the same page when you wrote _"why wear the media with a write?"_ I still haven't made sense of this question. :-)